Mocked exception type crashes junit runner and gradle exits with success

Previously reported on JIRA: <https://issues.gradle.org/browse/GRADLE-3556>

If a unit throws a mocked exception with a null stacktrace then the junit test runner fails to mark the test as failed and more importantly it incorrectly exits with code 0 even though the test runner itself encountered an error.

### Expected Behavior
The test should be marked as failed because an exception was not expected to be thrown:

```
:test

com.example.BuggerTest > thisTestShouldBeMarkedAsFailed FAILED
    $java.lang.RuntimeException$$EnhancerByMockitoWithCGLIB$$4bc6e699

1 test completed, 1 failed
:test FAILED

FAILURE: Build failed with an exception.
```


### Current Behavior
The test is not marked as failed and gradle exits with success:

```
:compileJava
:processResources NO-SOURCE
:classes
:compileTestJava
:processTestResources NO-SOURCE
:testClasses
:test
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/0:0:0:0:0:0:0:1:48260'.
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:90)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:263)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
        at java.lang.Throwable.setStackTrace(Throwable.java:864)
        at org.gradle.internal.serialize.ExceptionPlaceholder.read(ExceptionPlaceholder.java:154)
        at org.gradle.internal.serialize.TopLevelExceptionPlaceholder.read(TopLevelExceptionPlaceholder.java:24)
        at org.gradle.internal.serialize.ExceptionReplacingObjectInputStream.doResolveObject(ExceptionReplacingObjectInputStream.java:67)
        at org.gradle.internal.serialize.ExceptionReplacingObjectInputStream$1.transform(ExceptionReplacingObjectInputStream.java:31)
        at org.gradle.internal.serialize.ExceptionReplacingObjectInputStream.resolveObject(ExceptionReplacingObjectInputStream.java:62)
        at java.io.ObjectInputStream.checkResolve(ObjectInputStream.java:1583)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:422)
        at org.gradle.internal.serialize.Message.receive(Message.java:55)
        at org.gradle.internal.serialize.BaseSerializerFactory$ThrowableSerializer.read(BaseSerializerFactory.java:235)
        at org.gradle.internal.serialize.BaseSerializerFactory$ThrowableSerializer.read(BaseSerializerFactory.java:233)
        at org.gradle.internal.remote.internal.hub.DefaultMethodArgsSerializer$ArraySerializer.read(DefaultMethodArgsSerializer.java:71)
        at org.gradle.internal.remote.internal.hub.DefaultMethodArgsSerializer$ArraySerializer.read(DefaultMethodArgsSerializer.java:60)
        at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationReader.readArguments(MethodInvocationSerializer.java:124)
        at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationReader.read(MethodInvocationSerializer.java:119)
        at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationReader.read(MethodInvocationSerializer.java:99)
        at org.gradle.internal.serialize.kryo.TypeSafeSerializer$1.read(TypeSafeSerializer.java:34)
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:66)
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:79)
        ... 6 more
:check

BUILD SUCCESSFUL
```

### Context
Our unit tests verify that we handle exceptions properly. As we were migrating between library versions the expected exception had changed but our CI test runner did not report which test had failed because gradle exited with success. We only noticed any problem because the mutation tests failed due to "test not being green".

### Steps to Reproduce (for bugs)
I have created a minimum example project here: <https://github.com/spacecowboy/gradle-junit-bug>

There is a line you can uncomment to make the stacktrace "non-null".

### Your Environment
 * Build scan URL: <https://scans.gradle.com/s/e4gf2urqlgq7q>

--- a/subprojects/messaging/src/integTest/groovy/org/gradle/internal/serialize/ExceptionPlaceholderIntegrationTest.groovy
+++ b/subprojects/messaging/src/integTest/groovy/org/gradle/internal/serialize/ExceptionPlaceholderIntegrationTest.groovy
@@ -1,0 +1,68 @@
+/*
+ * Copyright 2012 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.serialize
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+class ExceptionPlaceholderIntegrationTest extends AbstractIntegrationSpec {
+
+    @Issue("https://github.com/gradle/gradle/issues/1618")
+    def "internal exception should not be thrown"() {
+        given:
+        buildFile << '''
+            apply plugin: 'java'
+
+            repositories {
+                jcenter()
+            }
+
+            dependencies {
+                testCompile 'junit:junit:4.12'
+                testCompile 'org.mockito:mockito-core:2.3.7'
+            }
+        '''
+
+        file('src/test/java/example/Issue1618Test.java') << '''
+        package example;
+
+        import org.junit.Test;
+
+        import static org.mockito.Mockito.doThrow;
+        import static org.mockito.Mockito.mock;
+
+        public class Issue1618Test {
+
+            public static class Bugger {
+                public void run() {
+                }
+            }
+
+            @Test(expected = RuntimeException.class)
+            public void thisTestShouldBeMarkedAsFailed() {
+                RuntimeException mockedException = mock(RuntimeException.class);
+                Bugger bugger = mock(Bugger.class);
+                doThrow(mockedException).when(bugger).run();
+                bugger.run();
+            }
+        }
+        '''
+
+        expect:
+        succeeds('test')
+    }
+}

--- a/subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java
+++ b/subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java
@@ -45,7 +45,7 @@ class ExceptionPlaceholder implements Serializable {
         type = throwable.getClass().getName();
 
         try {
-            stackTrace = throwable.getStackTrace();
+            stackTrace = throwable.getStackTrace() == null ? new StackTraceElement[0] : throwable.getStackTrace();
         } catch (Throwable ignored) {
 // TODO:ADAM - switch the logging back on. Need to make sending messages from daemon to client async wrt log event generation
 //                LOGGER.debug("Ignoring failure to extract throwable stack trace.", ignored);

--- a/subprojects/messaging/src/test/groovy/org/gradle/internal/serialize/MessageTest.groovy
+++ b/subprojects/messaging/src/test/groovy/org/gradle/internal/serialize/MessageTest.groovy
@@ -259,6 +259,18 @@ class MessageTest extends Specification {
         e2.message == 'broken toString()'
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/1618")
+    def "transports mock exception with null cause"() {
+        def exceptionWithNullCause = Mock(Exception)
+        exceptionWithNullCause.stackTrace >> null
+
+        when:
+        def transported = transport(exceptionWithNullCause)
+
+        then:
+        transported.stackTrace == [] as StackTraceElement[]
+    }
+
     @Issue("GRADLE-1996")
     def "can transport exception that implements writeReplace()"() {
         def original = new WriteReplaceException("original")

